### PR TITLE
Bookmarkview

### PR DESCRIPTION
**Pull request by albert-smit**
for issue #150

### CODE REVIEW

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/annotation/controller/BookmarkController.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+    /**
+     * Initialize the shout listeners.
+     */
+    // java doesn't let us typecheck generics so the casts are unchecked
+    // checkstyle doesn't like assert without parentheses with generics
+    @SuppressWarnings({"checkstyle:genericwhitespace", "unchecked"})
+    private void initListeners() {
+        listen(Message.BOOKMARKS, (sender, subject, args) -> {
+            show();
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/annotation/controller/BookmarkController.java]: redundant curly brackets

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/annotation/controller/BookmarkController.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        listItems = FXCollections.observableArrayList();
+        filteredItems = new FilteredList<AbstractBookmark>(listItems);
+        bookmarkList.setItems(filteredItems);
+
+        /*
+         * create a new Cell factory so we can use a different method
+         * to create the text of each cell,
+         * and assign an event handler for mouse clicks.
+         */
+        bookmarkList.setCellFactory(new Callback<ListView<AbstractBookmark>,
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/annotation/controller/BookmarkController.java]: the inner anonymous class for `Callback` can be replaced by a lambda to make the code more readable and less bulky.

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/annotation/controller/BookmarkController.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    // checkstyle doesn't like assert without parentheses with generics
+    @SuppressWarnings({"checkstyle:genericwhitespace", "unchecked"})
+    private void initListeners() {
+        listen(Message.BOOKMARKS, (sender, subject, args) -> {
+            show();
+        });
+
+        listen(Message.LOADED, (sender, subject, args) -> {
+
+            switch (subject) {
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/annotation/controller/BookmarkController.java]: Redundant code in the cases, you should only set the list item to the appropriate value inside the case body.

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/annotation/controller/BookmarkController.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+            }
+           return annotation.toCellString().contains(searchBox.getCharacters());
+        });
+    }
+
+    /**
+     * makes the sidebar invisible.
+     */
+    private void hide() {
+        wrapper.visibleProperty().set(false);
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/annotation/controller/BookmarkController.java]: `wrapper.visibleProperty().set(...)` is duplicaate code, it should be moved (or changed) to a method that takes a boolean as argument

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java [hunk 5]
```diff
@@ -269,6 +275,14 @@ private void initListeners() {
                     repaintNow = true;
                     repaint();
                 });
+
+        listen(Message.GOTO, (controller, subject, args) -> {
+            assert args[0] instanceof Long;
+            Long position = ((Long) args[0]);
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java]: Redundant parentheses

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java [hunk 6]
```diff
@@ -481,8 +497,7 @@ private void repaintPosition(final double position) {
         if (zoomLevel > zoomSwitchLevel) {
             if (currentZoomLevel != zoomLevel || repaintNow) {
                 Group diagramDrawing = new Group();
-                double width = getMaxUnifiedEnd(graph) * scale
-                        * VertexView.HORIZONTALSCALE;
+                double width = unifiedEnd * scale * VertexView.HORIZONTALSCALE;
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java]: Duplicate code; replace  `scale * VertexView.HORIZONTALSCALE` with `scaledVertex`

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java [hunk 7]
```diff
@@ -506,7 +521,7 @@ private void repaintPosition(final double position) {
                 graphDrawing.getChildren().add(
                         drawGraph(startBucket, endBucket));
                 graphDrawing.getChildren().add(
-                        new Rectangle(getMaxUnifiedEnd(graph) * scale
+                        new Rectangle(unifiedEnd * scale
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java]: Duplicate code; replace  `scale * VertexView.HORIZONTALSCALE` with `scaledVertex`

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java [hunk 8]
```diff
@@ -481,8 +497,7 @@ private void repaintPosition(final double position) {
         if (zoomLevel > zoomSwitchLevel) {
             if (currentZoomLevel != zoomLevel || repaintNow) {
                 Group diagramDrawing = new Group();
-                double width = getMaxUnifiedEnd(graph) * scale
-                        * VertexView.HORIZONTALSCALE;
+                double width = unifiedEnd * scale * VertexView.HORIZONTALSCALE;
                 int diagramLevel = zoomLevel - zoomSwitchLevel;
                 diagramDrawing.getChildren().add(
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java]: `graphDrawing.getChildren()` is duplicate code

File: lifetiles-launcher/src/main/resources/fxml/BookmarkView.fxml [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+			</right>
+		</BorderPane>	
+		<HBox>
+			<children>
+				<TextField fx:id="searchBox" promptText="Search ..." 
+					onAction="#searchAction" HBox.hgrow="ALWAYS"/>
+				<Button text="Search" onMouseClicked="#searchAction" />
+			</children>
+		</HBox>		
+		<Separator ></Separator>
```
Comment (jorenham) [inline lifetiles-launcher/src/main/resources/fxml/BookmarkView.fxml]: Could be replaced with `<Separator />`

File: lifetiles-launcher/src/main/resources/fxml/BookmarkView.fxml [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+		<ListView fx:id="bookmarkList" VBox.vgrow="ALWAYS"></ListView>
```
Comment (jorenham) [inline lifetiles-launcher/src/main/resources/fxml/BookmarkView.fxml]: Could be replaced with `<ListView fx:id="bookmarkList" VBox.vgrow="ALWAYS" />`

File: lifetiles-launcher/src/main/resources/fxml/BookmarkView.fxml [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+<?import javafx.scene.layout.BorderPane?>
+
+<VBox xmlns:fx="http://javafx.com/fxml/1" minHeight="0.0"
+	minWidth="0.0" prefHeight="720" prefWidth="300.0" maxWidth="500.0"
+	fx:controller="nl.tudelft.lifetiles.annotation.controller.BookmarkController"
+	fx:id="wrapper">
+	<children>
+		<BorderPane>
+			<left>
+				<Label text="Bookmarks and annotations"></Label>
```
Comment (jorenham) [inline lifetiles-launcher/src/main/resources/fxml/BookmarkView.fxml]: Could be replaced with `<Label text="Bookmarks and annotations" />`

File: lifetiles-launcher/src/main/resources/styles/BookmarkStyle.css [hunk 12]
```diff
@@ -0,0 +1,24 @@
+/* Stylesheet for the Bookmark view. */
```
Comment (jorenham) [inline lifetiles-launcher/src/main/resources/styles/BookmarkStyle.css:1]: This file needs a reformat

**Comment by jorenham** [comment]
Moar tests!

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java [hunk 13]
```diff
@@ -91,9 +92,13 @@
...
 
     /**
-     * The view controller.
+     * graph model.
      */
     private Graph<SequenceSegment> graph;
+    /**
+     * the highest unified coordinate in the graph.
+     */
+    private long unifiedEnd;
```
Comment (rutgervandenberg) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java]: Confusing variable name. Should make it clear that it is the highest end.

**Comment by rutgervandenberg** [comment]
Looks good.

**Comment by alangerak** [comment]
Looks good when above comments are implemented.

**Pull request merged**